src/__tests__/transforms.test.ts

```diff
@@ -39,8 +54,28 @@ describe("ensure", () => {
       [p => p.age > p.name.length, "unexpected age"],
     );
 
-    expect(personDecoder({ name: "Foo", age: 32 })).toEqual(error("name is too far from perfect"));
-    expect(personDecoder({ name: "IevgenFoo", age: 5 })).toEqual(error("unexpected age"));
+    expect(personDecoder({ name: "Foo", age: 32 })).toEqual({
+      type: "error",
+      value: [
+        {
+          path: [],
+          received: { name: "Foo", age: 32 },
+          error: "name is too far from perfect",
+        },
+      ],
+    });
+    expect(personDecoder({ name: "IevgenFoo", age: 5 })).toEqual(
+      error([
+        {
+          path: [],
+          received: {
+            name: "IevgenFoo",
+            age: 5,
+          },
+          error: "unexpected age",
+        },
+      ]),
+    );
```

Comment:
It would be nice to have a list of errors here in response to the object  that violates both conditions:
{ name: "Foo", age: 32 }

In this case, we will have 
-the same path
-the same value in all 'received' fields of the array.

We can leave this duplication as is 
OR
replace error with errors:[] (need a notion of non-empty array) and group all errors with the same path (this may be nice for the user)

What do you think ?

(it a major API change, lets better spend more time on thinking but fo it right)

Reply:
This looks like a flat-or-not-flat problem or a group-or-not-group problem that deserves some careful consideration. Ultimately, I think it depends on how the library is used - I personally tend to fix one decoder error at a time, in which case I can just go one-by-one in order until all errors are out. In this case, additional nesting doesn't help and I'm just annoyed by having to expand a more deeply nested objects. Others might develop a different workflow and be annoyed by other things.

I propose we sit on this for a bit and if the change still makes sense a while from now, then let's do it.

Reply:
+